api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java

```diff
@@ -0,0 +1,139 @@
+  }
+
+  /**
+   * checkClusters deletes older clusters in order to force an upgrade on the
+   * next researcher login. This method is meant to be restricted to invocation
+   * by App Engine cron.
+   *
+   * The cluster deletion policy here aims to strike a balance between enforcing
+   * upgrades, cost savings, and minimizing user disruption. To this point, our
+   * goal is to only upgrade idle clusters if possible, as doing so gives us an
+   * assurance that a researcher is not actively using it. We delete clusters in
+   * the following cases:
+   *
+   * 1. It exceeds the max cluster age. Per environment, but O(weeks).
+   * 2. Since a cluster was last accessed, its spent over half its remaining
+   *    lifespan idle (as defined by [1]). This dynamic limit means that as the
+   *    upgrade horizon approaches, we'll be more aggressive in upgrading.
+   */
+  @Override
+  public ResponseEntity<CheckClustersResponse> checkClusters() {
+    WorkbenchConfig config = configProvider.get();
+    Instant now = clock.instant();
+    Duration maxAge = Duration.ofDays(config.firecloud.clusterMaxAgeDays);
+
+    ClusterApi clusterApi = clusterApiProvider.get();
+    List<Cluster> clusters;
+    try {
+      clusters = clusterApi.listClusters(null, false);
+    } catch (ApiException e) {
+      throw ExceptionUtils.convertNotebookException(e);
+    }
+
+    int errors = 0;
+    int activeDeletes = 0;
+    int unusedDeletes = 0;
+    for (Cluster c : clusters) {
+      String clusterId = c.getGoogleProject() + "/" + c.getClusterName();
+      if (ClusterStatus.UNKNOWN.equals(c.getStatus()) || c.getStatus() == null) {
+        log.warning(String.format("unknown cluster status for cluster '%s'", clusterId));
+        continue;
+      }
+      if (!ClusterStatus.RUNNING.equals(c.getStatus()) &&
+          !ClusterStatus.STOPPED.equals(c.getStatus())) {
+        // For now, we only handle running or stopped (suspended) clusters.
+        continue;
+      }
+
+      Instant created = Instant.parse(c.getCreatedDate());
+      Duration age = Duration.between(created, now);
+      Duration lifeRemaining = maxAge.minus(age);
+      Instant lastUsed = Instant.parse(c.getDateAccessed());
+      Duration unusedDuration = Duration.between(lastUsed, now);
+      if (lifeRemaining.isNegative()) {
+        log.info(String.format(
+            "deleting cluster '%s', exceeded max lifetime @ %s (>%s)",
+            clusterId, formatDuration(age), formatDuration(maxAge)));
+        activeDeletes++;
+      } else if (lifeRemaining.toMillis() < unusedDuration.toMillis()) {
+        log.info(String.format(
+            "deleting cluster '%s', unused for %s, more than its remaining lifespan (>%s)",
+            clusterId, formatDuration(unusedDuration), formatDuration(lifeRemaining)));
+        unusedDeletes++;
+      } else {
+        // Don't delete.
+        continue;
+      }
+      try {
+        clusterApi.deleteCluster(c.getGoogleProject(), c.getClusterName());
```

Comment:
There's a race condition here: we fetch the big list up top, start iterating and deleting clusters, and by the time we get to the deleteCluster call for a cluster, it's actually in use again. So we may want to add a check here before deleting a cluster to make sure it's still not running. (Even better would be to have deleteCluster provide the option to say "don't delete if I'm in use"... but I assume that doesn't exist.)

Reply:
This comment was written assuming we would attempt to *not* delete clusters that were running, or at least as aggressively as we'd delete paused clusters... I misread the logic initially. See above.

Reply:
Added. Ultimately this is just mitigation as we can't do this transactionally (without Leo API support) - this is a matter of reducing the window from 10s of seconds -> 1s of seconds.